cmd/common/registry.go

```diff
@@ -0,0 +1,377 @@
+// Copyright 2022-2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/frontends/console"
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/utils"
+	cols "github.com/inspektor-gadget/inspektor-gadget/pkg/columns"
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	gadgetregistry "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-registry"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/parser"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+)
+
+// AddCommandsFromRegistry adds all gadgets known by the registry as cobra commands as a subcommand to their categories
+func AddCommandsFromRegistry(rootCmd *cobra.Command, runtime runtime.Runtime, columnFilters []cols.ColumnFilter) {
+	runtimeParams := runtime.GlobalParamDescs().ToParams()
+
+	// Build lookup
+	lookup := make(map[string]*cobra.Command)
+	for _, cmd := range rootCmd.Commands() {
+		lookup[cmd.Name()] = cmd
```

Comment:
What's the reasoning here? To avoid replacing a category command if this was filled previously? (Do we need / want that behavior)

Reply:
This was built to support mixing the old method (manually adding the categories/commands) with the new method when we planned to migrate the gadgets separately - so I think we can remove this now.